File: sample/ts/11_lissajous_particles.ts
// このファイルは EAST ベース TypeScript プレビュー出力です。
// TODO: 専用 TSEmitter 実装へ段階移行する。
import * as math from "./math.js";
import { perf_counter } from "./time.js";
import { save_gif } from "./pytra/runtime/gif.js";

// 11: Sample that outputs Lissajous-motion particles as a GIF.

function color_palette() {
    let p = bytearray();
    for (let i = 0; i < 256; i += 1) {
        let r = i;
        let g = i * 3 % 256;
        let b = 255 - i;
        p.push(r);
        p.push(g);
        p.push(b);
    }
    return bytes(p);
}

function run_11_lissajous_particles() {
    let w = 320;
    let h = 240;
    let frames_n = 360;
    let particles = 48;
    let out_path = "sample/out/11_lissajous_particles.gif";
    
    let start = perf_counter();
    let frames = [];
    
    for (let t = 0; t < frames_n; t += 1) {
        let frame = bytearray(w * h);
        
        for (let p = 0; p < particles; p += 1) {
            let phase = p * 0.261799;
            let x = Math.trunc(Number(w * 0.5 + w * 0.38 * math.sin(0.11 * t + phase * 2.0)));
            let y = Math.trunc(Number(h * 0.5 + h * 0.38 * math.sin(0.17 * t + phase * 3.0)));
            let color = 30 + p * 9 % 220;
            
            for (let dy = -2; dy < 3; dy += 1) {
                for (let dx = -2; dx < 3; dx += 1) {
                    let xx = x + dx;
                    let yy = y + dy;
                    if (xx >= 0 && xx < w && yy >= 0 && yy < h) {
                        let d2 = dx * dx + dy * dy;
                        if (d2 <= 4) {
                            let idx = yy * w + xx;
                            let v = color - d2 * 20;
                            v = max(0, v);
                            if (v > frame[idx]) {
                                frame[idx] = v;
                            }
                        }
                    }
                }
            }
        }
        frames.push(bytes(frame));
    }
    save_gif(out_path, w, h, frames, color_palette());
    let elapsed = perf_counter() - start;
    console.log("output:", out_path);
    console.log("frames:", frames_n);
    console.log("elapsed_sec:", elapsed);
}

// __main__ guard
run_11_lissajous_particles();

File: sample/cpp/11_lissajous_particles.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/built_in/numeric_ops.h"
#include "generated/std/math.h"
#include "generated/std/time.h"
#include "generated/utils/gif.h"

// 11: Sample that outputs Lissajous-motion particles as a GIF.

bytes color_palette() {
    bytearray p = bytearray{};
    for (int64 i = 0; i < 256; ++i) {
        int64 r = i;
        int64 g = i * 3 % 256;
        int64 b = 255 - i;
        p.append(static_cast<uint8>(static_cast<int64>(r)));
        p.append(static_cast<uint8>(static_cast<int64>(g)));
        p.append(static_cast<uint8>(static_cast<int64>(b)));
    }
    return p;
}

void run_11_lissajous_particles() {
    int64 w = 320;
    int64 h = 240;
    int64 frames_n = 360;
    int64 particles = 48;
    str out_path = "sample/out/11_lissajous_particles.gif";
    
    float64 start = pytra::std::time::perf_counter();
    rc<list<bytes>> frames = rc_list_from_value(list<bytes>{});
    
    for (int64 t = 0; t < frames_n; ++t) {
        bytearray frame = bytearray(w * h);
        float64 __hoisted_cast_1 = static_cast<float64>(t);
        
        for (int64 p = 0; p < particles; ++p) {
            float64 phase = static_cast<float64>(p) * 0.261799;
            int64 x = static_cast<int64>(static_cast<float64>(w) * 0.5 + static_cast<float64>(w) * 0.38 * pytra::std::math::sin(0.11 * __hoisted_cast_1 + phase * 2.0));
            int64 y = static_cast<int64>(static_cast<float64>(h) * 0.5 + static_cast<float64>(h) * 0.38 * pytra::std::math::sin(0.17 * __hoisted_cast_1 + phase * 3.0));
            int64 color = 30 + p * 9 % 220;
            
            for (int64 dy = -(2); dy < 3; ++dy) {
                for (int64 dx = -(2); dx < 3; ++dx) {
                    int64 xx = x + dx;
                    int64 yy = y + dy;
                    if ((xx >= 0) && (xx < w) && (yy >= 0) && (yy < h)) {
                        int64 d2 = dx * dx + dy * dy;
                        if (d2 <= 4) {
                            int64 idx = yy * w + xx;
                            int64 v = color - d2 * 20;
                            v = ::std::max<int64>(static_cast<int64>(0), static_cast<int64>(v));
                            if (v > frame[idx])
                                frame[idx] = uint8(static_cast<int64>(v));
                        }
                    }
                }
            }
        }
        rc_list_ref(frames).append(frame);
    }
    pytra::utils::gif::save_gif(out_path, w, h, rc_list_ref(frames), color_palette(), 3, 0);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", frames_n);
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_11_lissajous_particles();
    return 0;
}
